x/ecocredit/server/core/bridge_receive.go

```diff
@@ -0,0 +1,138 @@
+package core
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	api "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+)
+
+// BridgeReceive bridges credits received from another chain.
+func (k Keeper) BridgeReceive(ctx context.Context, req *core.MsgBridgeReceive) (*core.MsgBridgeReceiveResponse, error) {
+	bridgeServiceAddr, err := sdk.AccAddressFromBech32(req.Issuer)
+	if err != nil {
+		return nil, err
+	}
+
+	class, err := k.stateStore.ClassTable().GetById(ctx, req.Project.ClassId)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("could not get class with id %s: %s", req.Project.ClassId, err.Error())
+	}
+
+	// first we check if there is an existing project
+	idx := api.ProjectClassKeyReferenceIdIndexKey{}.WithClassKeyReferenceId(class.Key, req.Project.ReferenceId)
+	it, err := k.stateStore.ProjectTable().List(ctx, idx)
+	if err != nil {
+		return nil, err
+	}
+
+	// we only want the first project that matches the reference ID, so we do not loop here.
+	var project *api.Project
+	if it.Next() {
+		var err error
+		project, err = it.Value()
+		if err != nil {
+			return nil, err
+		}
+	}
+	it.Close()
+
+	// if no project was found, create one + issue batch
+	if project == nil {
+		projectRes, err := k.CreateProject(ctx, &core.MsgCreateProject{
+			Admin:        req.Issuer,
+			ClassId:      req.Project.ClassId,
+			Metadata:     req.Project.Metadata,
+			Jurisdiction: req.Project.Jurisdiction,
+			ReferenceId:  req.Project.ReferenceId,
+		})
+		if err != nil {
+			return nil, err
+		}
+		batchRes, err := k.CreateBatch(ctx, &core.MsgCreateBatch{
+			Issuer:    req.Issuer,
+			ProjectId: projectRes.ProjectId,
+			Issuance: []*core.BatchIssuance{
+				{Recipient: req.Batch.Recipient, TradableAmount: req.Batch.Amount},
+			},
+			Metadata:  req.Batch.Metadata,
+			StartDate: req.Batch.StartDate,
+			EndDate:   req.Batch.EndDate,
+			Open:      true,
+			OriginTx:  req.Batch.OriginTx,
+			Note:      req.Batch.Note,
+		})
+		if err != nil {
+			return nil, err
+		}
+		return &core.MsgBridgeReceiveResponse{BatchDenom: batchRes.BatchDenom, ProjectId: projectRes.ProjectId}, nil
+	}
```

Comment:
Should we be using `else` and returning the response once at the end of `BridgeReceive`. I'm noticing we don't have an event for `BridgeReceive` and we may want to add one and we would need to add to each condition.

If each condition was handled without the returned response, we could define the event and response before running through the conditions, set within each condition, and then emit and return at the end.

Reply:
We can add an event in a follow up and consider a single return response when/if we do. Ok to leave as is but was having a hard time following each condition, which might also be resolved by pulling out the matching logic.

Reply:
hows this https://github.com/regen-network/regen-ledger/pull/1174/commits/794e070947ce135fa8739adeb7d82b84abd24f74 ?

Reply:
This helps :+1: